runner/src/server/plugins/engine/pluginHandlers/exit/index.ts

```diff
@@ -0,0 +1,105 @@
+import { HapiRequest, HapiResponseToolkit } from "server/types";
+import Boom from "boom";
+import { getForm } from "./prehandlers/getForm";
+import { parseExitEmailErrors } from "./prehandlers/parseExitEmailErrors";
+import { getState } from "./prehandlers/getState";
+import { getBacklink } from "./prehandlers/getBacklink";
+import { validateEmailPostRequest } from "./prehandlers/validateEmailPostRequest";
+import { checkUserIsAllowedAccess } from "./prehandlers/checkUserIsAllowedAccess";
+
+export const emailGet = {
+  method: "get",
+  path: "/{id}/exit/email",
+  options: {
+    description: "Allows users to exit the form if enabled on the form",
+    pre: [
+      {
+        assign: "form",
+        method: getForm,
+      },
+      {
+        assign: "state",
+        method: getState,
+      },
+      { method: checkUserIsAllowedAccess },
+      {
+        assign: "errors",
+        method: parseExitEmailErrors,
+      },
+      {
+        assign: "backlink",
+        method: getBacklink,
+      },
+    ],
+  },
+  handler: async (request: HapiRequest, h: HapiResponseToolkit) => {
+    return h.view("exit/email", {
+      errors: request.pre.errors,
+      backLink: request.pre.backlink,
+    });
+  },
+};
+
+export const emailPost = {
+  method: "post",
+  path: "/{id}/exit/email",
+  options: {
+    pre: [
+      {
+        assign: "form",
+        method: getForm,
+      },
+      {
+        assign: "state",
+        method: getState,
+      },
+      { method: checkUserIsAllowedAccess },
+      {
+        method: validateEmailPostRequest,
+      },
+    ],
+  },
+  handler: async (request: HapiRequest, h: HapiResponseToolkit) => {
+    const form = request.pre.form;
+
+    const { cacheService, exitService } = request.services([]);
+    const state = await cacheService.getState(request);
+
+    try {
+      await exitService.exitForm(form, state);
```

Comment:
Does the returned data from `exitService.exitForm` need to be kept and passed through in the redirect? I just did a test passing back the expiry date and it looked like `exitForm` handled it properly, but the variable didn't make it to the exit status template.

Reply:
Good catch! I suppose that would've been caught if I'd written the smoke tests :).

Yeah that should be saved to the session. (When I was drafting the feature I was just using req.yar to store the response, then added `cacheService.setExitState` but didn't fully migrate things across)